### PR TITLE
Add setting `SQLITE_JOURNAL_MODE` to enable WAL

### PR DESCRIPTION
**Pull request by noerw**
[WAL journal mode](https://www.sqlite.org/wal.html) is supported in SQLite and provides much better concurrent write performance than the default mode. This likely enables even large instances to get along with a sqlite DB (though I didn't do any load testing yet — does someone have a load testing setup ready?).

After empirically confirming WAL performs better, it could be made the default setting, but installations with builds for non-unix & -windows systems may break (see disclaimer in the linked sqlite doc). Not sure if thats a real-world issue we need to consider, as gitea only provides builds for unix & windows afaik.

### CODE REVIEW

**Review by zeripath — APPROVED** [review]
Tbh I'd be happy if we just changed it to wal by default or even provided an os detection that made the decision to use wal by default based on os

**Comment by noerw** [comment]
@zeripath  Sounds reasonable. Would you prefer choosing the appropriate mode during runtime or build time?

**Comment by noerw** [comment]
- I made it select the mode via a runtime OS check. Would be nice if someone could check if WAL works on the BSDs.

- I also removed the setting, as the automatic selection should be good enough. If you think the setting is still needed, feel free to revert 1f744f9e0a8efa2209489b5682046d13bd362fb8

- The default journal mode `DELETE` could be changed to `TRUNCATE`, which [provides better performance](https://www.sqlite.org/pragma.html#pragma_journal_mode) and has no drawbacks apart from the journal file lingering around:
  > The TRUNCATE journaling mode commits transactions by truncating the rollback journal to zero-length instead of deleting it. On many systems, truncating a file is much faster than deleting the file since the containing directory does not need to be changed.

**Comment by wxiaoguang** [comment]
I'd like to trust the SQLite document and think that SQLite WAL works on all modern OS's native filesystem.

However, according to the document, SQLite's Shared-Memory WAL doesn't work for filesystems without shared-memory support.

There is a section for non-Shared-Memory WAL:  "8. Use of WAL Without Shared-Memory"

So IMO the problem is not related to OS, but related to filesystem and how the program uses SQLite.

**Comment by zeripath** [comment]
Damn, sorry I was too hasty and misremembered what SQLite says - I should have rechecked the docs. 

We should restore the option. (and likely back to your original commit - see below.)

---

Now, what to do about the default/blank setting of that option. 

Unfortunately, I guess we can't set it to `WAL` because that could actually be a breaking change and I would not be surprised if it would break some people's installs. Docker can use "network filesystems" internally (especially docker for windows) so they're already banned from `WAL`.

We could do the soft-default option - add an option, default it to blank, but at install time provide it as an option with a default set to `WAL`. But the problem is how to detect if it is failing, if SQLite just barfs at open time that's a simple test. (It would of course be best if it just fell back to `DELETE`.) If it doesn't barf on open, I worry it could just quietly corrupt itself or barf only on the first write which could lead to some very confusing issues. (you can check if sqlite has quietly fallen back to `DELETE` by opening the db with `sqlite3` and running `PRAGMA schema_mode;`)

Of course if SQLite just auto falls back to `DELETE` when it detects that `WAL` would not work then all of my worries are irrelevant...

I guess the simplest thing to do is to just add the option but leave it blank. 

Otherwise we'll need to test common configurations and see if we can get things to barf. If I had to guess at a likely failure configuration it would be docker for windows. 

PS @noerw `runtime.GOOS` is a constant so any code dependent on _should_ be resolved at compile time.

**Comment by noerw** [comment]
I just did a quick test, running a build of this current PR with a `PRAGMA journal_mode=DELETE;` db on a cifs samba mount.
Sqlite happily enabled WAL mode there.
As this is not supported according to the sqlite docs, enabling WAL and falling back (explicitly or implicitly within sqlite) is not an option without elaborate detection of the underlying OS & filesystem. That doesn't feel reasonable.

So I think the way to go is to revert to the original version of this PR

**Comment by zeripath** [comment]
So thinking on I think we should not set the `_journal_mode` if the `SQLITE_JOURNAL_MODE` has not been explicitly provided.

This is because clever users might have already changed to use WAL in their dbs and you never know sqlite might change their default in future.

```patch
From 3fd7c2886a8b43f0528c06a28fc99a20ffd8e4d4 Mon Sep 17 00:00:00 2001
From: Andrew Thornton <art27@cantab.net>
Date: Sat, 30 Jul 2022 13:34:26 +0100
Subject: [PATCH] Do not set _journal_mode if `SQLITE_JOURNAL_MODE` is unset

Signed-off-by: Andrew Thornton <art27@cantab.net>

diff --git a/modules/setting/database.go b/modules/setting/database.go
index 0b37826f2..6c241d4d2 100644
--- a/modules/setting/database.go
+++ b/modules/setting/database.go
@@ -92,7 +92,7 @@ func InitDBConfig() {
 
 	Database.Path = sec.Key("PATH").MustString(filepath.Join(AppDataPath, "gitea.db"))
 	Database.Timeout = sec.Key("SQLITE_TIMEOUT").MustInt(500)
-	Database.SQliteJournalMode = sec.Key("SQLITE_JOURNAL_MODE").MustString("DELETE")
+	Database.SQliteJournalMode = sec.Key("SQLITE_JOURNAL_MODE").MustString("")
 
 	Database.MaxIdleConns = sec.Key("MAX_IDLE_CONNS").MustInt(2)
 	if Database.UseMySQL {
@@ -139,8 +139,12 @@ func DBConnStr() (string, error) {
 		if err := os.MkdirAll(path.Dir(Database.Path), os.ModePerm); err != nil {
 			return "", fmt.Errorf("Failed to create directories: %v", err)
 		}
-		connStr = fmt.Sprintf("file:%s?cache=shared&mode=rwc&_busy_timeout=%d&_txlock=immediate&_journal_mode=%s",
-			Database.Path, Database.Timeout, Database.SQliteJournalMode)
+		journalMode := ""
+		if Database.SQliteJournalMode != "" {
+			journalMode = "&_journal_mode=" + Database.SQliteJournalMode
+		}
+		connStr = fmt.Sprintf("file:%s?cache=shared&mode=rwc&_busy_timeout=%d&_txlock=immediate%s",
+			Database.Path, Database.Timeout, journalMode)
 	default:
 		return "", fmt.Errorf("Unknown database type: %s", Database.Type)
 	}
-- 
2.37.1

```

File: modules/setting/database.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ var (
 		LogSQL            bool
 		Charset           string
 		Timeout           int // seconds
+		SQliteJournalMode string
 		UseSQLite3        bool
```
Comment (wxiaoguang) [inline modules/setting/database.go]: `SQliteJournalMode`, the `SQlite` seems not the same as the word in `UseSQLite3` below.

Maybe `SQLiteJournalMode`?

**Review by wxiaoguang — APPROVED** [review]

**Review by zeripath — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
.

**Comment by 6543** [comment]
.

**Pull request merged**